xbmc/guilib/GUITextLayout.cpp

```diff
@@ -579,63 +579,117 @@ void CGUITextLayout::WrapText(const vecText &text, float maxWidth)
+    auto widthOf = [&](const std::vector<character_t>::const_iterator start,
+                       const std::vector<character_t>::const_iterator end) -> float
+    {
+      if (start == end)
+        return 0.0f;
+      tmp.assign(start, end);
+      return m_font->GetTextWidth(tmp);
+    };
 
-    while (pos < line.m_text.end())
+    auto skipLeadingSpaces = [&](vecText::const_iterator& it)
     {
-      // Get the current letter in the string
-      const character_t& letter = *pos;
+      while (it != line.m_text.end() && CanWrapAtLetter(*it))
+        ++it;
+    };
+
+    auto current = line.m_text.begin();
+    skipLeadingSpaces(current);
 
-      if (CanWrapAtLetter(letter)) // Check for a space char
-        lastSpacePos = pos;
+    float currentWidth = 0.0f;
+    std::vector<character_t>::const_iterator currentStart = current;
+    // track end without trailing spaces
+    std::vector<character_t>::const_iterator lastNonSpaceInLine = currentStart;
 
-      curLine.emplace_back(letter);
+    while (current != line.m_text.end())
+    {
+      // Find next candidate wrap position
+      std::vector<character_t>::const_iterator wordEnd = current;
+      while (wordEnd != line.m_text.end() && !CanWrapAtLetter(*wordEnd))
+        ++wordEnd;
 
-      const float currWidth = m_font->GetTextWidth(curLine);
+      const bool hasSpace = (wordEnd != line.m_text.end());
+      const float wordWidth = widthOf(current, wordEnd);
+      const float spaceWidth = hasSpace ? widthOf(wordEnd, wordEnd + 1) : 0.0f;
 
-      if (currWidth > maxWidth)
+      // Try to include word + trailing space
+      if (currentWidth + wordWidth + spaceWidth <= maxWidth)
       {
-        if (lastSpacePos > pos) // No space char where split the line, so split by char
-        {
-          // If the pos is equal to lastBeginPos, maxWidth is not large enough to contain 1 character
-          // Push a line with the single character and move on to the next character.
-          if (pos == lastBeginPos)
-            ++pos;
-
-          CGUIString linePart{lastBeginPos, pos, false};
-          m_lines.emplace_back(linePart);
-        }
-        else
-        {
-          CGUIString linePart{lastBeginPos, lastSpacePos, false};
-          m_lines.emplace_back(linePart);
+        currentWidth += wordWidth + spaceWidth;
+        lastNonSpaceInLine = wordEnd; // exclude trailing space
+        current = wordEnd;
+        if (hasSpace)
+          ++current;
+        continue;
+      }
 
-          pos = lastSpacePos + 1;
-          lastSpacePos = line.m_text.end();
-        }
+      // Try to include word without trailing space
+      if (currentWidth + wordWidth <= maxWidth)
+      {
+        m_lines.emplace_back(currentStart, wordEnd, false);
+        if (m_lines.size() >= nMaxLines)
+          return;
 
-        curLine.clear();
-        lastBeginPos = pos;
+        current = hasSpace ? (wordEnd + 1) : wordEnd;
+        skipLeadingSpaces(current);
```

Comment:
Is the line 636 necessary? As we are at the end of a word isn't the next char a space and handled by `skipLeadingSpaces` anyway?